### DISCUSSION TITLE
Feature suggestion: Option to create a backup of the last played game at GBM startup.

### DISCUSSION THREAD

**Discussion by NewRavenGer** [MikeMaximus/gbm#220] [Ideas]
Hi there, it's me again :) 
I don't know how feasible and most important how much of general use such an option would be. However, after thinking quite a bit about other solutions for my small problem - maybe with third-party-tools or scripts - I couldn't come up with anything really useful, reliable and working.

The problem: my girlfriend usually uses her PC in a remote way... gaming on her laptop via remote play, gaming on tv (and sometimes even on her android tablet) via Steam Link. The issue here is... she also shuts down the host (-> her PC) with these remote devices. In some - not that many - cases it happens, that Windows is already shutting down while GBM hasn't completed the backup, especially if there are  either larger savegames files (Minecraft, in our example) or many of them (Skyrim, in our example) so there is a lot of data to compress. I already tried to change the compression settings to "fast" and even store/none but that doesn't reliable help (and no, telling her "don't use that option, just walk the stairs up and turn of your PC like everyone else" didn't either). 

So, an option to simply just do a backup of the single last played game as soon as GBM starts (which would be when the PC is turned on again) would solve my issue here. So I would like to ask if something like that is possible to implement at some point.

(another option could be some kind of scheduler to create backups in fixed time intervals, but it seems a bit of bloat for GBM and maybe is not worth the required work, so therefor I think, the above option would be more suitable)

**Comment by MikeMaximus** [comment]
I'm pretty sure I can just delay the Windows shut down until a backup operation is completed, that will be the cleanest way to do it.

This can be done by refusing to let GBM close if there is an operation in progress, which it currently doesn't do.

**Comment by NewRavenGer** [comment]
Thank you. Well, if this could solve the issue, that would be fine by me of course. 
However, I may add that something similar also came to my mind and I have already tried to start tools which should prevent a shutdown while running (some sync- and backup-tools) and none of them actually was able to prevent that behavior. Seems to me like Valve would use some fairly harsh methods to shut down the host system (acts something similar to shutdown -f). But even if I'm wrong here, as far as I know, the "WaitToKill"-TimeOut in Windows is, per default, 20 seconds which isn't enough for longer backup tasks. Not sure if it is even possible to delay the closing of an application any longer than that at all. On the other hand... usually you come up with even better solutions to my issues than I had, so I'm fairly sure this time it's no different.

**Comment by MikeMaximus** [comment]
The reason I think it can be done is because during the development of an older version, I once had a bug in the shutdown code that seemed to cause GBM to delay Windows shutdown indefinitely.  It would just sit on the waiting to shutdown screen until shutdown was forced.

But as you say Valve might be using a non-standard method of shutdown that may invalidate what I was doing to cause that behavior.

**Comment by NewRavenGer** [comment]
Well, if you want, I would gladly try a test build to see if it's working. I would actually prefer your "cleaner" way to do this myself, but if it's not working, we can still return to my "backup after restart"-idea, maybe.

**Comment by MikeMaximus** [comment]
Yeah definitely, that's what I was thinking,  I'll post a link to a build in this issue to try once I get around to working on it.

**Comment by MikeMaximus** [comment]
I've created a very simple test, this build will attempt to prevent Windows from shutting down in general while GBM is running.  If this works in your scenarios as well I'll proceed with the rest of the logic.

Backup your current GBM.exe and replace it with the one in the archive. 

This is based on the current master of v1.2.5, it won't cause any issues going back to the release version.

GBM - Prevent Shutdown Build (x64) EDIT: Link Removed
Build: 2021-02-20T23:43:47
GBM.exe SHA256: 70D188C5F9C2E68F05A823B0732284C5FA48EABB296AB24607D267F783111E5B

**Comment by NewRavenGer** [comment]
Thank you. Nope, doesn't seem to do anything here... I tried a few options:

1. just having GBM in the background, doing nothing, trying to shut down the OS via start menu. PC shuts down.
2. having GBM in the background creating a (large) backup for a game and trying to shut down the OS via start menu. PC shuts down
3. with GBM in background, having a game running long enough to trigger a backup in Steams Big Picture, closed the game and try to shut down the PC with the Steam Big Picture menu. PC shuts down.

(4. with GBM in background, having a game running in Steams Big Picture and try to shut down the PC without even closing the game. PC shuts down.)

Well, 4 hasn't anything to do with GBM, but I was not amused to see that Steam even actually offers to shut down windows without quitting the game.

However, none of the shutdowns were prevented and they didn't even seem to take any longer. I will try one the second PC now, but that probably will take a few minutes. I will edit the results in here :)

Edit: same results on the other PC.

**Comment by MikeMaximus** [comment]
Hmm I just tested it again to make sure I packaged the right executable.  It prevented shut down or sign out then timed out back to Windows after a period of time. 

What version of Windows are you using?

**Comment by NewRavenGer** [comment]
Nope, nothing of that ever happened here. It didn't even show that "waiting for applications to close" screen which you sometimes get when programs are still running while you try to shut down or restart in the first 3 cases (it did show that screen a short moment on test 4, but hadn't even listed GBM there, only Steam itself and the not closed game). Is it that screen where it "stuck" for you before it went back after the time-out? Because if it was, I may have an idea why it didn't work for me...

This would be the German version of Windows 10 20H2 (19042) on both machines.

**Comment by MikeMaximus** [comment]
I discovered why it's not working, it only prevents shutdown if the main window is shown and not hidden in the tray.  It doesn't need focus, it just needs to not be hidden.  I'm assuming this is some kind of prevention mechanism for people that are coding malware to take advantage of this, but I need to dig a bit more.

**Comment by NewRavenGer** [comment]
Yeah, figured out the same thing probably at the same time as you did :) Plus... you should absolutely not have Windows to AutoEndTasks via registry setting. So, with the main window open and reverting that Windows setting, I could reproduce what this new version should do. 

The "waiting" screen was there for about 20 seconds, then I got logged out. After I logged in again, GBM was still active, the window was still open, however, the backup process failed, but I guess that was to be expected (probably because there wasn't a user logged in and the backup couldn't be finished in these 20 seconds)

**Comment by MikeMaximus** [comment]
Hmm I wonder if it's because the backup process is actually a spawned 7za.exe and that IS being closed.   I wasn't counting on that. 😞 

I think we will need to go with your idea.  I was thinking that if GBM detects a shutdown call while a backup is running,  it can just auto-queue it for next start up.

**Comment by NewRavenGer** [comment]
Seems that way. However, if GBM has started that 7za child process (which is has and windows also registered it that way, according to process explorer), shouldn't it stay active as long as GBM.exe is active (unless GBM closed it or it finished it's job, of course)?

But maybe that restart-thingy isn't even that bad. You could probably already set-up that "backup on the next start-up"-task  right after the backup is triggered and right before the actual backup starts. If it finishes without issue, you just could remove that "task" at the end again. This would probably be quite transparent, it would not add any duplicate backups that aren't needed, and you wouldn't even need to add any options for it. On the other hand, I can clearly see why you aren't that happy with such a solution, but I ran out of ideas how to solve this otherwise myself, too, so I probably won't be much of a help.

**Comment by MikeMaximus** [comment]
Yes you'd think that's how it would work since 7za is a child process of GBM, but I can't see any other reason GBM would indicate that the backup failed once the 20 seconds was up.  Do you recall the exact wording in the log?  Did 7za show an error or did the output just end?  If GBM states "operation failed due to an error" means that 7za shut down unexpectedly.

Thanks for the ideas, I'll think on it a bit more.  It is something i'd like to solve for you.

**Comment by NewRavenGer** [comment]
Well, I redid the whole thing and that's the (complete) log:

> [21.02.2021 14:21:38] A sync from the master game list has been triggered.
> [21.02.2021 14:21:39] 0 change(s) synced.
> [21.02.2021 14:21:39] Game List (182) Loaded.
> [21.02.2021 14:21:56] A sync to the master game list has been triggered.
> [21.02.2021 14:21:56] 0 change(s) synced.
> [21.02.2021 14:21:56] Game List (182) Loaded.
> [21.02.2021 14:21:56] Adding The Elder Scrolls IV: Oblivion to backup queue.
> [21.02.2021 14:21:57] The Elder Scrolls IV: Oblivion Save Folder Size: 7,75 GB
> [21.02.2021 14:21:57] Backup of D:\Documents\My Games\Oblivion\Saves in progress...

However, above the log window it says:
"Last Action: The Elder Scrolls IV: Oblivion backup failed at 2:22 (sic!) ." which is correct (even if the 12-hour timestamp only used in this spot seems kind of strange), because there isn't any updated backup archive in the backup directory. If you want me to, I could check if there is something in the temp/working directories, but I'm fairly sure there is. It just wasn't able to complete.

**Comment by MikeMaximus** [comment]
Thanks for checking again,  I tried my own test as well, it's kind of frustrating that I can't use the debugger while testing this scenario. 😀 

I'm fairly certain now that in addition to killing child processes it's also killing all other threads except the main UI thread, which is why the backup output simply ends and the main thread reports a failure.  I think these are security measures in .NET so you can't pull any malware type shenanigans during shutdown.

I'm going to work on another solution,  I can't give you a time estimate, but I'll update this issue when/if there's anything you can test.

**Comment by MikeMaximus** [comment]
I want to keep feature requests in the discussion section now, so I've converted this issue.

**Reply by NewRavenGer** [reply]
That's fine by me. As far as testing goes... I'll just wait patiently for any news (and thankfully, I also get notifications anyway) and I test everything you throw at me as soon as I see it ;) Thanks again for your work on this... and on GBM in general.

**Comment by MikeMaximus** [comment]
I think the following solution will solve this issue and also cover other situations such as power outages.

The backup queue will now be stored (local db) and each configuration in queue is removed as the backup is completed.  On startup, if there's anything left in the queue it will be resumed automatically.  This failsafe queue is wiped when operations are manually canceled or anytime a new backup queue is generated.

Test build is below, based on the [v126](https://github.com/MikeMaximus/gbm/tree/v126) branch.

[GBM v1.2.6 Beta 2 (x64)](https://www.dropbox.com/s/bopf7m0dye3fmzc/gbm.v1.2.6.b2.7z?dl=0)
Build: 2021-02-28T16:44:58
GBM.exe SHA256: B131622CDC7EBC6F853A41A72B7F378D157A64E5E49CB7A2EE117F25EC78F521

**Since the beta contains database updates, if you revert back to v1.2.5 you should also restore your gbm.s3db.v125.bak files in your backup folder and in %LOCALAPPDATA%\gbm.**

Let me know how it goes if you get a chance to test it.  😃

**Reply by NewRavenGer** [reply]
Well, I was trying to try, but I only got a message about the database error, then followed by an unhandled exception ( https://imgur.com/a/vHBz2yk ).  :)

**Reply by MikeMaximus** [reply]
Sorry!  I copy pasted the db upgrade code into the wrong place and since mine was already upgraded I didn't catch it.  🤦

I'll edit the original post in a few minutes with a new build.

**Comment by NewRavenGer** [comment]
So far so good. Worked fine here with a manual triggered backup and with a backup started through monitoring. I maybe would consider adding a small delay here (60 seconds or something like that) before the actual backup process starts, because when GBM starts with windows, there is usually a fairly high load, but it's working fine so far. 

Small question, because I hadn't the time to experiment with it, yet: does the adding to the queue happen 

a) when the game starts 
b) when the game starts, but after the set "short session" time
or c) when the game got closed?

I guess it's b), right?

**Reply by MikeMaximus** [reply]
It would be c), the backup queue is only generated at the time the game is closed.  So this only works when GBM is actually preparing to or in the act of performing the backup at the time it's unexpectedly closed.  It won't be queued if the game is detected but still being played at the time it's closed.

I see what you mean about delaying the operation because of window startup, but there may be some issues with that.  I'll look into it.

**Reply by NewRavenGer** [reply]
Yeah, after thinking about it a bit more: a power outage potentially could also result in savegames not being correctly written to disk by the game itself, and you probably don't want to back up these then. Guess it makes more sense that  (c-)way, yeah.

Well, if this may produce issues, I guess there is always the option to delay the start of the whole program on the user side, so maybe it's all fine the way it is.

**Reply by MikeMaximus** [reply]
Yes a scheduled startup task with a delay would accomplish the same thing.  I'll still take a look next time I'm working on it.

**Reply by MikeMaximus** [reply]
I decided not to implement a delay on the start up tasks.  The task scheduler gives much finer control over startup anyway, such as the exact time of the delay, etc.  EDIT:  Turns out "much finer control" is a bit of an exaggeration as the delay time options it gives you in the UI are pretty sparse. 😆

I'm sure you already got that figured out, but I did write a [guide](http://mikemaximus.github.io/gbm-web/gbm_task_scheduler.html) on how to use the task scheduler to start GBM as Admin.  I'll add another section for setting a start-up delay as well for anyone come across this and needs help.

**Reply by NewRavenGer** [reply]
That's perfectly fine by me. 

However... in that regard, wouldn't it be a smart idea to actually use the task scheduler in general to do the autostart of GBM? I mean, I do it this way since like... forever anyway, but that guide would probably a lot easier to understand for "not that advanced" users if the profile would already exist. And since GBM has an autostart-option anyway but using the registry to do it, maybe a change here would be beneficial (that being said, this way you could probably also implement the delay if you want to, without interfering with GBMs inner workings at all). Of course, this is just a suggestion, but I don't see any obvious reason not to do it this way.

However, what I really need to point out at this point, again, is, how much I really - and I mean really - appreciate your support. You listen to every request, suggestion or report, you really help out even when someone comes here and talks about probably fairly rare use cases (like this one here), and you still invest your time and try to make it work. Thank you!

**Reply by MikeMaximus** [reply]
You're welcome!  I always try to help out if possible, I'm happy that people are using the app and finding it useful.

The reason I don't use the task scheduler for GBM startup by default is because .NET doesn't provide a nice way to interface with it.  From what I understand you need to use Win32 calls to interface with it, a .NET wrapper or schtasks.exe.  It just seemed like a lot of extra work and bloat for the benefit at the time.  I may take another look into the wrapper for it.

**Reply by NewRavenGer** [reply]
Well... because it IS useful, and it became soooo versatile with the complete dynamic paths, the chaining of profiles and so on. You could easily use it for a lot more things than savegame-backup if you want to. I still have some small ideas which I would like to see at some point, but damn, when I found GBM about 3 years ago and rather cautious tried it to replace my savegame-releated backup-solution back then, I never thought GBM would become so powerful in such a short amount of time. 

I see, after reading a bit into creating a scheduled task using vb.net, I would agree. While certainly doable, it seems like the needed work outweighs the benefits indeed. So, yeah... I guess we could consider this "feature suggestion" done :)

**Reply by MikeMaximus** [reply]
Yes I think we can consider this one done.  If you'd like to continue using the beta version it should be fine until the release version is ready.  This version will likely be a smaller maintenance release,  just waiting to see if anymore issues with the new UI pop up.

**Reply by NewRavenGer** [reply]
Well, I won't be able to help with testing that part, because... well... uhm... I have disabled as much as I could from it, to be honest 😓
However, as long as I don't run into some serious issues with the beta, I guess I will use it on my own machine for now, but wait for the release on the other machines here. Thanks, again 😀

**Reply by MikeMaximus** [reply]
Oh that's fine, not everyone likes UI changes, it's reason I added the options to disable.  It actually still helps with testing as there could be weird bugs specific to having everything disabled. 😁

**Reply by NewRavenGer** [reply]
Well, it's not that I don't like it, it looks nice, it probably works great, but I just don't see me using the game launcher at all and I like my backup-software the most the less I see of it. So, for me, all the UI changes were either redundant, because I have the required stuff in the game manager or just not of much use. And therefor all the stuff is gone... 😅  I was actually a bit sad at first that I now have to deal with tabs and buttons and such stuff in my beloved Game Manager and can't see everything at one glance anymore (especially the monitor options are too important to be hidden under a button, imo). However, even I have to agree that it looks a lot more modern, sorted and less cluttered now and offers just more space to implement more possible options in the future. So I'm totally fine with it, actually. It's just nothing that is beneficial in my use case.

**Comment by NewRavenGer** [comment]
I actually found a small issue with the 3.26 Beta. Everytime a game isn't closed regulary but instead crashes, I get an unhandled exception/access denied error. This didn't happen with 3.25 plus, after I restart GBM it sometimes crashes again (maybe because of the queue?). However, the same games backup successful if they are closed regularly, and I've never had the issue with any earlier version. Maybe this helps:

************** Ausnahmetext **************
System.ComponentModel.Win32Exception (0x80004005): Zugriff verweigert
 bei System.Diagnostics.ProcessManager.OpenProcess(Int32 processId, Int32 access, Boolean throwIfExited)
 bei System.Diagnostics.NtProcessManager.GetModuleInfos(Int32 processId, Boolean firstModuleOnly)
 bei System.Diagnostics.NtProcessManager.GetFirstModuleInfo(Int32 processId)
 bei System.Diagnostics.Process.get_MainModule()
 bei GBM.frmMain.SetGameInfo(Boolean bMulti)
 bei GBM.frmMain.ScanTimerEventProcessor(Object myObject, EventArgs myEventArgs)
 bei System.Windows.Forms.Timer.OnTick(EventArgs e)
 bei System.Windows.Forms.Timer.TimerNativeWindow.WndProc(Message& m)
 bei System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
************** Geladene Assemblys **************
mscorlib
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4300.0 built by: NET48REL1LAST_C.
 CodeBase: file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dll.
----------------------------------------
GBM
 Assembly-Version: 1.2.6.20359.
 Win32-Version: 1.2.6.0.
 CodeBase: file:///D:/GameBackupMonitor/GBM.exe.
----------------------------------------
Microsoft.VisualBasic
 Assembly-Version: 10.0.0.0.
 Win32-Version: 14.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/Microsoft.VisualBasic/v4.0_10.0.0.0__b03f5f7f11d50a3a/Microsoft.VisualBasic.dll.
----------------------------------------
System
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4300.0 built by: NET48REL1LAST_C.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll.
----------------------------------------
System.Core
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4320.0 built by: NET48REL1LAST_C.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll.
----------------------------------------
System.Windows.Forms
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4270.0 built by: NET48REL1LAST_C.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll.
----------------------------------------
System.Drawing
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll.
----------------------------------------
System.Configuration
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4190.0 built by: NET48REL1LAST_B.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll.
----------------------------------------
System.Xml
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll.
----------------------------------------
System.Runtime.Remoting
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Runtime.Remoting/v4.0_4.0.0.0__b77a5c561934e089/System.Runtime.Remoting.dll.
----------------------------------------
System.Data
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4270.0 built by: NET48REL1LAST_C.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_64/System.Data/v4.0_4.0.0.0__b77a5c561934e089/System.Data.dll.
----------------------------------------
Mono.Data.Sqlite
 Assembly-Version: 4.0.0.0.
 Win32-Version: 1.0.61.0.
 CodeBase: file:///D:/GameBackupMonitor/Mono.Data.Sqlite.DLL.
----------------------------------------
System.Transactions
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_64/System.Transactions/v4.0_4.0.0.0__b77a5c561934e089/System.Transactions.dll.
----------------------------------------
System.EnterpriseServices
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_64/System.EnterpriseServices/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.EnterpriseServices.dll.
----------------------------------------
System.Numerics
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Numerics/v4.0_4.0.0.0__b77a5c561934e089/System.Numerics.dll.
----------------------------------------
System.Management
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Management/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Management.dll.
----------------------------------------
Accessibility
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/Accessibility/v4.0_4.0.0.0__b03f5f7f11d50a3a/Accessibility.dll.
----------------------------------------
Microsoft.GeneratedCode
 Assembly-Version: 1.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll.
----------------------------------------
System.Windows.Forms.resources
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms.resources/v4.0_4.0.0.0_de_b77a5c561934e089/System.Windows.Forms.resources.dll.
----------------------------------------
mscorlib.resources
 Assembly-Version: 4.0.0.0.
 Win32-Version: 4.8.4084.0 built by: NET48REL1.
 CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/mscorlib.resources/v4.0_4.0.0.0_de_b77a5c561934e089/mscorlib.resources.dll.
----------------------------------------

And sorry, GitHub screws up the formatting no matter if I use the code-tag in this case or not.

**Reply by MikeMaximus** [reply]
Strange, the area it's crashing didn't receive any changes.  I'll take a look at what could be causing that.

**Reply by NewRavenGer** [reply]
Thank you :) If you need some further details, I'll help as good as I can.

**Reply by MikeMaximus** [reply]
I believe I see the problem indicated by this error message, but it should also have been happening in v1.2.5.  I'm not sure why it would crash again on restart either.  Was the crash on restart a different error message?

**Reply by NewRavenGer** [reply]
No, I only glanced over it shortly, but both were access denied errors for sure. However, I can always try to reproduce it to check this out - this may, however, take a while because usually both games where this happened don't crash for me. Because of that, there is a possibility that it would have also happened in 1.25 but just the situation to do so never happened. This is also something that I could try, I still have a backup of my 1.25 database.

**Reply by MikeMaximus** [reply]
If you can reproduce the crash on opening the app again that would help.  I'm pretty sure I fixed what was causing GBM to get the access denied crash after the game crashed, it's related to the new icon handling I put in v1.2.5.  

I'm just unsure why it would crash again on opening, as it shouldn't be running any code at that point to cause an access denied exception that I can think of.

**Reply by MikeMaximus** [reply]
I misinterpreted your initial post about the error message that happened again after you restarted GBM.  They likely were exactly the same and I don't want you to waste your time on trying to reproduce it.  I updated the beta build and it would be better if you just used that instead to see if the same crash occurs with those particular games.

[GBM v1.2.6 Beta 2 (x64)](https://www.dropbox.com/s/bopf7m0dye3fmzc/gbm.v1.2.6.b2.7z?dl=0)
Build: 2021-02-28T16:44:58
GBM.exe SHA256: B131622CDC7EBC6F853A41A72B7F378D157A64E5E49CB7A2EE117F25EC78F521

The problem here was that sometimes GBM can't access the details of a detected process, and when this happens it'll throw an access denied exception.  When I reworked the code that grabs the icon from the process it wasn't included in the exception handling, so it was just crashing instead of simply reporting it in the log.

There's many reasons this access denied exception can get thrown.   These are just some examples: 
- When you're running a game as admin and GBM isn't running as admin.
- The game has some kind of anti-tamper/cheat that prevents reading the module.
- The game isn't fully initialized or crashes at the time GBM detects it and tries to read the module.

This crash may actually be pretty annoying for many people due to the variety of ways it can happen, so thank you for reporting it!

**Reply by NewRavenGer** [reply]
Thank you, I will update my version and post again should this happen again. :)

Yeah, it actually is kind of hard trying to get such a crash "on purpose". The first game actually never crashed for me before, the second one only sometimes in LAN co-op (but of course not, when I want it to crash). Probably would have taken a few days, I guess.

**Comment by NewRavenGer** [comment]
### Edit: actually, kind of ignore that for now. After restarting everything, I couldn't get GBM to do anything like that again and since I had the server of the game also running, maybe I messed something up here. 


Nah, there still seems to be something off, somehow... while not an exception-error-message and no crash, I now have literary about 50 entries like that generated in the last 5 minutes (these are only the last few lines) in the log. This time, there wasn't any crash of the game or something like that. The game was normally closed, there isn't any related process of it running in the background and the savegame-backup-file was actually created fine, however... after that, GBM still tries to do... something. Over and over again.

[01.03.2021 20:55:13] Valheim backup was cancelled due to session length.
[01.03.2021 20:55:23] Valheim Detected
[01.03.2021 20:55:23] An error occured while obtaining the details of the detected game.
Error: Zugriff verweigert
[01.03.2021 20:55:23] Valheim has ended.
[01.03.2021 20:55:23] A sync to the master game list has been triggered.
[01.03.2021 20:55:24] 0 change(s) synced.
[01.03.2021 20:55:24] An error has occured while saving the icon for Valheim:

Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
[01.03.2021 20:55:24] Valheim backup was cancelled due to session length.
[01.03.2021 20:55:34] Valheim Detected
[01.03.2021 20:55:34] An error occured while obtaining the details of the detected game.
Error: Zugriff verweigert
[01.03.2021 20:55:34] Valheim has ended.
[01.03.2021 20:55:34] A sync to the master game list has been triggered.
[01.03.2021 20:55:34] 0 change(s) synced.
[01.03.2021 20:55:34] An error has occured while saving the icon for Valheim:

Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
[01.03.2021 20:55:34] Valheim backup was cancelled due to session length.

- Zugriff verweigert -> access denied
- Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt. -> The object reference was not set to an object instance.

This was with Beta2, of course. :)

**Reply by MikeMaximus** [reply]
There definitely is still some kind of bug here with the new icon caching, I'm currently trying to replicate it myself with no luck.  Is the Valheim server process also called Valheim.exe?  If both the client and server share the same executable name, make sure the Game Path in the GBM configuration is pointed at the client location and not the server.

I've determined what is causing the "The object reference was not set to an object instance." error message when trying to save the icon and can fix that easily.  But the issue with it doing that over and over again may be related to your Valheim server and/or client.

**Reply by NewRavenGer** [reply]
No, the server is called valheim_server.exe (and I don't use wildcards or regex in my profiles), however, I can't rule out for sure that I didn't screw up somewhere between all the starting the server, starting the game, restarting the server, restarting the game. Even if I didn't do it nearly as much 50 times and from the timestamps in the log, I was already done with that testing stuff at this point. However, after a fresh restart, I can't reproduce anything similar to that stuff.

**Reply by MikeMaximus** [reply]
I noticed Valheim crashes on exit sometimes, maybe a copy of it was doing something weird in the background in combination with the Unity error reporter.

**Reply by NewRavenGer** [reply]
That is certainly possible... however, even then... the log states various times "Valheim has ended." before it says "Valheim Detected" about 10 seconds later. If there was some weird background stuff happening, to trigger GBM that way (and especially triggering it that often) wouldn't this require a constant restart of the game process for something like this to happen? However, I'll keep an eye on this, for sure, but for now, everything seems fine.

**Reply by MikeMaximus** [reply]
Yes it would require the game to be constantly starting and ending, that's why I was theorizing the unity error handler may have been doing something very odd to cause that.

Thanks for taking the time to report and help me diagnose these issues, I noticed and fixed 3 other small bugs as well while diagnosing this one.

Now back to building my Barley farm on the Plains of Valheim.  😁

**Reply by MikeMaximus** [reply]
I've been looking through the logic in GBM to see how it could get into a detection loop like that, since it's still bothering me and I've never seen it do that before over the years I've been working on this.

I think that there was a stale Valheim process stuck running in the background, such as one that crashed but never fully closed.  GBM would see this process still running every 10 seconds (this is GBM's default detection pass speed), but it was throwing an exception trying to lock onto it to use WaitForExit.  The code handles this part of detection just ignores the exception and tries again in 10 seconds, which would be the source of the repeating detection and ending loop.  Situations like that should be pretty rare, but the code that handles this was written like 6 years ago and it's sloppy.  I'm going to do a bit of tweaking to make sure that can't happen anymore.

I just thought you'd like to know, since my theory blaming the game itself or unity error handling was pretty weak. 🤦

**Reply by NewRavenGer** [reply]
Nah, it was alright, because there probably was "something" still in the background. I mean, it has to from a logical standpoint. I remember a similar situation a few years back when I had freshly discovered GBM. While modding Oblivion back then I tried many things out, had file injectors like OBSE running, removed some of them, tried other ones. GBM did nearly the same thing at some point... however, with two differences: back then I clearly could see Oblivion.exe still lurking in the background - that wasn't the case with Valheim, at least not anymore at that point I noticed GBM was acting "strange". And of course, back then GBM was only stating "Oblivion detected" followed by a "Oblivion backup was cancelled due to session length" in a loop, without any other error messages. But as I understand, this is related to the icon generating thing you already fixed. 

For my understanding - besides that error messages, of course, GBM did exactly what it should do. Detect the game process and do its thing. GBM can't really know (and shouldn't try to know) if the process is running because the game is still active and played or if it's still there because of a weird stuck-half-crash-whatever-issue. While I didn't see Valheim.exe in the task manager, I'm fairly sure it was kind of there... somehow. I just was too fast with typing here before double-checking that stuff, which I absolutely should have done.

**Reply by MikeMaximus** [reply]
It does sound like the problem with Oblivion was the same thing, this issue would have been around since the early beta versions of GBM.

It's true that GBM was just doing what it was meant to do, but I've still improved the error handling in that section so instead completely ignoring an unexpected error and looping when this occurs, it should catch it, display it in the log and stop detection so the user can deal with it.
